### PR TITLE
feat: Add locator types supported by flutter integration driver

### PR DESCRIPTION
**Pull request by sudharsan-selvaraj**
## Change list

Please provide briefly described change list which are you going to propose. 
 
## Types of changes

What types of changes are you proposing/introducing to Java client?
_Put an `x` in the boxes that apply_

- [ ] No changes in production code.
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Details

Please provide more details about changes if it is necessary. If there are new features you can provide code samples which show the way they
work and possible use cases. Also you can create [gists](https://gist.github.com) with pasted java code samples or put them here using markdown. 
About markdown please read [Mastering markdown](https://guides.github.com/features/mastering-markdown/) and [Writing on GitHub](https://help.github.com/categories/writing-on-github/)

### CODE REVIEW

File: src/main/java/io/appium/java_client/AppiumBy.java [hunk 1]
```diff
@@ -53,6 +56,13 @@ protected AppiumBy(String selector, String locatorString, String locatorName) {
         return String.format("%s.%s: %s", AppiumBy.class.getSimpleName(), locatorName, remoteParameters.value());
     }
 
+    public Map<String, Object> toJson() {
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/AppiumBy.java:59]: not sure this method is needed

File: src/main/java/io/appium/java_client/AppiumBy.java [hunk 2]
```diff
@@ -257,4 +312,34 @@ protected ByIosNsPredicate(String locatorString) {
             super("-ios predicate string", locatorString, "iOSNsPredicate");
         }
     }
+
+    public static class ByFlutterType extends AppiumBy implements Serializable {
+        protected ByFlutterType(String locatorString) {
+            super("-flutter type", locatorString, "flutterType");
```
[the remark this answers is not in the document]
Comment (sudharsan-selvaraj) [reply src/main/java/io/appium/java_client/AppiumBy.java:318]: Yeah, I have made the required changes to both driver and server package.

https://github.com/AppiumTestDistribution/appium-flutter-integration-driver/commit/42cf91463d48ee6cad0dec542a79c95e98eee8fa

File: src/main/java/io/appium/java_client/AppiumBy.java (hunk 1 above)
Comment (mykola-mokhnach) [reply src/main/java/io/appium/java_client/AppiumBy.java:59]: I can observe that Remotable.Parameters class already contains `toJson` method. It's private though. Lets add it to the list in https://github.com/SeleniumHQ/selenium/issues/13949 and leave a TODO there.

Also consider using Map.of() factory instead.

Comment (mykola-mokhnach) [reply src/main/java/io/appium/java_client/AppiumBy.java:59]: Could you also provide an example of any of the above calls where toJson method may be used?

Comment (sudharsan-selvaraj) [reply src/main/java/io/appium/java_client/AppiumBy.java:59]: One such example is using `flutter: scrollTillVisible` method.

1. **Without toJson()**

```
driver.executeScript("flutter: scrollTillVisible", Map.of("finder", Map.of("using", "-flutter key", "value", "login_button")));
```

2. **With toJson()**
```
driver.executeScript("flutter: scrollTillVisible", Map.of("finder", AppiumBy.flutterKey("login_button").toJson());
```

Comment (mykola-mokhnach) [reply src/main/java/io/appium/java_client/AppiumBy.java:59]: It would probably make sense to create FlutterIosDriver and FlutterAndroidDriver inherited from iOSDriver and AndroidDriver, so the above extensions may be implemented as first-class methods, e.g.

flutter. scrollTillVisible(AppiumBy.flutterKey("login_button"))

Comment (sudharsan-selvaraj) [reply src/main/java/io/appium/java_client/AppiumBy.java:59]: We initially had a similar idea. After reviewing the implementation, we found that the logic is consistent for both iOS and Android drivers, so we decided to maintain this approach.

Comment (saikrishna321) [reply src/main/java/io/appium/java_client/AppiumBy.java:59]: @sudharsan-selvaraj Agree we wanted to make it consistent but nothing is stopping us implementing the initial approach if everyone is fine. It gets easier to implement the tests.
cc: @mykola-mokhnach

Comment (sudharsan-selvaraj) [reply src/main/java/io/appium/java_client/AppiumBy.java:59]: @saikrishna321 @mykola-mokhnach  That sounds good! Let's proceed by creating separate Flutter-specific driver classes to abstract the underlying commands. I'll prepare a patch with the changes and tag for the review.

File: src/main/java/io/appium/java_client/flutter/FlutterAndroidDriver.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+import org.openqa.selenium.remote.HttpCommandExecutor;
+import org.openqa.selenium.remote.http.ClientConfig;
+import org.openqa.selenium.remote.http.HttpClient;
+
+import java.net.URL;
+
+/**
+ * Custom AndroidDriver implementation with additional Flutter-specific capabilities.
+ */
+public class FlutterAndroidDriver extends AndroidDriver implements FlutterDriver {
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/flutter/FlutterAndroidDriver.java:23]: Please do not combine everything into a single big PR. Lets do locators, android and ios drivers each in a separate PRs

Comment (sudharsan-selvaraj) [reply src/main/java/io/appium/java_client/flutter/FlutterAndroidDriver.java:23]: Sure, Will scope this PR only to locator changes and move out platform specific driver implementation to different change set.

File: src/main/java/io/appium/java_client/flutter/FlutterAndroidDriver.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    /**
+     * Waits for an element to become visible on the Flutter application.
+     *
+     * @param option WaitOptions specifying the element and optional timeout.
+     */
+    @Override
+    public void waitForVisible(WaitOptions option) {
+        CommonFlutterCommands.waitForVisible(this, option);
```
Comment (SrinivasanTarget) [inline src/main/java/io/appium/java_client/flutter/FlutterAndroidDriver.java:82]: Agree on @mykola-mokhnach comments all these new driver part can be separate PR. And on your next PR it can be just FlutterCommands instead of CommonFlutterCommands.

File: src/main/java/io/appium/java_client/flutter/FlutterAndroidDriver.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        CommonFlutterCommands.scrollTillVisible(this, option);
+    }
+
+    /**
+     * Performs a double click on an element in the Flutter application.
+     *
+     * @param option DoubleClickOptions specifying the element to double-click.
+     */
+    @Override
+    public void performDoubleClick(DoubleClickOptions option) {
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/flutter/FlutterAndroidDriver.java:111]: Consider group commands to interfaces with default methods. later these interfaces could be inherited by both android and ios drivers to avoid unnecessary duplication

**Comment by sudharsan-selvaraj** [comment]
@mykola-mokhnach @SrinivasanTarget I have updated the PR to just include the changes related to locators. I'll create separate PR for driver implementation along with the e2e tests.

File: src/main/java/io/appium/java_client/AppiumBy.java [hunk 6]
```diff
@@ -26,13 +26,16 @@
 
 import java.io.Serializable;
 import java.util.List;
+import java.util.Map;
+
```
Comment (SrinivasanTarget) [inline src/main/java/io/appium/java_client/AppiumBy.java:30]: Extra spacing

File: src/main/java/io/appium/java_client/AppiumBy.java [hunk 7]
```diff
@@ -257,4 +323,56 @@ protected ByIosNsPredicate(String locatorString) {
             super("-ios predicate string", locatorString, "iOSNsPredicate");
         }
     }
+
+    public static class FlutterBy extends AppiumBy {
```
Comment (SrinivasanTarget) [inline src/main/java/io/appium/java_client/AppiumBy.java:327]: Can it be made abstract since its not going to be instantiated directly?

File: src/main/java/io/appium/java_client/AppiumBy.java [hunk 8]
```diff
@@ -257,4 +323,56 @@ protected ByIosNsPredicate(String locatorString) {
...
+            super(selector, locatorString, locatorName);
+        }
+
+        /**
+         * Parses an {@link AppiumBy.FlutterBy} object into a Map of parameters
+         * to be executed by flutter integration driver.
+         *
+         * @return A {@link Map} containing the parsed parameters
+         */
+        public Map<String, Object> toJson() {
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/AppiumBy.java:339]: Lets add this method with the driver, so it would be better visible where and how it is used

File: src/main/java/io/appium/java_client/AppiumBy.java (hunk 7 above)
Comment (sudharsan-selvaraj) [reply src/main/java/io/appium/java_client/AppiumBy.java:327]: Agree, will make the class abstract!

**Review by SrinivasanTarget — APPROVED** [review]

**Review by saikrishna321 — APPROVED** [review]

**Comment by jlipps** [comment]
Hi @sudharsan-selvaraj, congrats, the Appium project wants to compensate you for this (and perhaps other) contribution(s) this month! Please reply to this comment mentioning me and sharing your [OpenCollective](https://opencollective.com) account name, so that we can initiate payment! Or let me know if you decline to receive compensation via OpenCollective. Thank you!

**Comment by sudharsan-selvaraj** [comment]
Hey @jlipps , I would like to express my gratitude to the Appium team for the recognition and I'm pleased to be a part of the project. For the fund transfers, please use the OpenCollective account associated with https://opencollective.com/appium-device-farm. If you need any additional information, please let me know.

**Comment by maharaja-m** [comment]
Does this support annotated (eg @AndroidFindBy & @iOSXCUITFindBy) to create common element ?
